torchtune/modules/transforms/_transforms.py

```diff
@@ -0,0 +1,142 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import Any, List, Mapping, Protocol
+
+from torchtune.data import Message
+from torchtune.modules.tokenizers import ModelTokenizer
+
+
+class Transform(Protocol):
+    """
+    Loose interface for all data and model transforms. Transforms operate at the
+    sample level and perform operations on a sample dict which is contained in
+    kwargs. Any fields that will be processed are unfolded with explicit keyword-arguments,
+    then the updated dict is returned.
+    """
+
+    def __call__(self, **kwargs) -> Mapping[str, Any]:
+        pass
+
+
+class Compose(Transform):
+    """
+    Compose multiple transforms together, inspired by torchvision's ``Compose`` API
+
+    Args:
+        transforms (List[Transform]): List of transforms to compose together in sequential order.
+    """
+
+    def __init__(self, transforms: List[Transform]) -> None:
+        self.transforms = transforms
+
+    def __call__(self, **kwargs) -> Mapping[str, Any]:
+        for transform in self.transforms:
+            kwargs = transform(**kwargs)
+        return kwargs
+
+
+class TokenizeMessages(Transform):
+    """
+    Apply the ``tokenize_messages`` method from a given
+    :class:`~torchtune.modules.tokenizers.ModelTokenizer` on the ``messages`` field of the sample.
+
+    Args:
+        tokenizer (ModelTokenizer): Tokenizer used by the model that implements
+            the ``tokenize_messages`` method.
+    """
+
+    def __init__(self, tokenizer: ModelTokenizer, max_seq_len: int) -> None:
+        self.tokenizer = tokenizer
+        self.max_seq_len = max_seq_len
+
+    def __call__(self, *, messages: List[Message], **kwargs) -> Mapping[str, Any]:
+        tokens, mask = self.tokenizer.tokenize_messages(
+            messages, max_seq_len=self.max_seq_len
+        )
+        kwargs.update({"tokens": tokens, "mask": mask})
+        return kwargs
+
+
+class CrossAttentionMask(Transform):
+    """
+    Computes the cross-attention mask for text + image inputs. Text tokens that
+    participate in cross-attention with an image token will show True in the mask
+    and follow these rules:
+    1) Text tokens immediately following the image token up until the next image token
+    2) Consecutive image tokens attend to all subsequent text tokens
+
+    Resultant mask is of shape (text_seq_len, image_seq_len), where True indicates
+    that the token outputted from the image encoder attends to the token in the
+    text sequence.
+
+    Args:
+        num_patches (int): Number of patches per image, excluding class token.
```

Comment:
I wonder if we should add a link to modules/VisionTransformer for a in-depth explanation of what num_patches mean. For better clarity, would it make sense to rename it num_patches_per_tile, since later we multiply it by n_tiles?

If we say "number of patches per image", it may be confusing, because an image can have a variable number of patches.

later on you say:
```
single_image_seq_len = n_tiles * self.num_patches + 1
image_seq_len = single_image_seq_len * n_img
```

So tile != image. Image is a set of tiles.

Reply:
ah sorry, probably my shallow understanding of patches/images/tiles. What I intended was num_patches per tile. If it makes sense I'd like to keep the name consistent with your vision transformer (either patch_grid_size or patch_size maybe?), whichever parameter you use to compute num patches.

I also assumed at this point tiles is padded to the max in all the images. Is this incorrect? Where does the padding happen?

Reply:
Agreed, patch_size is good! I would just add to the documentation that its the number of patches **per tile**, instead of per image. In other words, whenever we use the word "image", i think it should mean the whole thing.

Reply:
patches_per_tile is a fixed size, and its calculated as (tile_size // patch_size)**2.

What I did in VisionTransformer was to ask the user to pass tile_size and patch_size, and I calculated it for them. The VisionTransformer has a helper function that saves this value: https://github.com/felipemello1/torchtune/blob/f683812626ad4559464840112ddce516487bea5c/torchtune/modules/vision_transformer.py#L249

Maybe get it form the model, or ask for tile_size and patch_size, to avoid user confusion?